File: Database_Integration/csharp/export/ExportFundsXml.cs
// =============================================================================
// EXPORT  —  relational database  ->  FundsXML file  (C# / .NET, SQLite).
//
// Standalone, copy-me example of ONE direction (DB -> FundsXML). The reverse
// is a separate project, ../import/. Over-commented as documentation.
//
// DB SCHEMA  ../../ddl/schema.sql  (already populated by ../import/).
//
// RUN
//   dotnet run --project Database_Integration/csharp/import -- fx.db some.xml
//   dotnet run --project Database_Integration/csharp/export -- \
//     fx.db FUNDSXML_MULTI_1 out.xml
//
// DEPENDENCIES  Microsoft.Data.Sqlite + System.Xml (BCL).
//
// FUNDSXML NOTES
//   * No XML namespace -> plain element names.
//   * xsi:noNamespaceSchemaLocation MUST be created in the XMLSchema-instance
//     namespace (a plain SetAttribute drops the prefix and a validator then
//     rejects it) -> CreateAttribute("xsi", ..., XsiNs).
//   * Constants the model does not store (TotalAssetNature=OFFICIAL, Price
//     ActionCode=C / PriceNature=OFFICIAL) reproduced verbatim so the
//     round-trip compares equal (../../tools/xml_equiv.py, always paired with
//     XSD validation).
//   * ORDER BY the 1-based *_seq columns reproduces the original order of
//     multiple funds / portfolios / positions.
// =============================================================================
using System;
using System.Globalization;
using System.IO;
using System.Xml;
using Microsoft.Data.Sqlite;

internal static class ExportFundsXml
{
    const string SchemaUrl =
        "https://github.com/fundsxml/schema/releases/download/4.2.9/FundsXML.xsd";
    const string XsiNs = "http://www.w3.org/2001/XMLSchema-instance";

    static readonly System.Collections.Generic.HashSet<string> PositionKinds =
        new() { "Equity", "Bond", "ShareClass", "Warrant", "Certificate",
                "Option", "Future", "FXForward", "Swap", "Repo", "RealEstate",
                "CallMoney", "Account", "Generic" };
    static readonly System.Collections.Generic.Dictionary<string, string> QtyElem =
        new() { ["Equity"] = "Units", ["Warrant"] = "Units",
                ["Certificate"] = "Units", ["Bond"] = "Nominal",
                ["ShareClass"] = "Shares", ["Option"] = "Contracts",
                ["Future"] = "Contracts" };

    static string Inv(double v) => v.ToString("0.00", CultureInfo.InvariantCulture);

    static XmlElement El(XmlDocument doc, XmlNode parent, string tag,
                         string? text = null)
    {
        var e = doc.CreateElement(tag);
        if (text != null) e.AppendChild(doc.CreateTextNode(text));
        parent.AppendChild(e);
        return e;
    }

    static SqliteDataReader Q(SqliteConnection c, string sql,
                              params object?[] ps)
    {
        var cmd = c.CreateCommand();
        cmd.CommandText = sql;
        for (int i = 0; i < ps.Length; i++)
            cmd.Parameters.AddWithValue($"@p{i}", ps[i] ?? DBNull.Value);
        return cmd.ExecuteReader();
    }
    static string? S(SqliteDataReader r, string col) =>
        r.IsDBNull(r.GetOrdinal(col)) ? null : r.GetString(r.GetOrdinal(col));
    static double D(SqliteDataReader r, string col) =>
        Convert.ToDouble(r.GetValue(r.GetOrdinal(col)),
            CultureInfo.InvariantCulture);
    static bool Null(SqliteDataReader r, string col) =>
        r.IsDBNull(r.GetOrdinal(col));

    static int Main(string[] args)
    {
        if (args.Length != 3)
        {
            Console.Error.WriteLine(
                "usage: ExportFundsXml <db> <document_id> <out.xml>");
            return 2;
        }
        string db = args[0], docId = args[1], outPath = args[2];

        using var c = new SqliteConnection($"Data Source={db}");
        c.Open();

        var doc = new XmlDocument();
        var root = doc.CreateElement("FundsXML4");
        // Must really be in the xsi namespace, else a validator rejects it.
        var sl = doc.CreateAttribute("xsi", "noNamespaceSchemaLocation", XsiNs);
        sl.Value = SchemaUrl;
        root.Attributes.Append(sl);
        doc.AppendChild(root);

        using (var d = Q(c, "SELECT * FROM document WHERE document_id=@p0",
                   docId))
        {
            if (!d.Read()) throw new Exception($"no document {docId}");
            var cd = El(doc, root, "ControlData");
            El(doc, cd, "UniqueDocumentID", S(d, "document_id"));
            // Regenerate timestamp; xml_equiv.py ignores its value.
            El(doc, cd, "DocumentGenerated",
                S(d, "generated") ?? "2025-10-02T00:00:00");
            if (!Null(d, "version"))                 // none for 4.0.0
                El(doc, cd, "Version", S(d, "version"));
            El(doc, cd, "ContentDate", S(d, "content_date"));
            var ds = El(doc, cd, "DataSupplier");
            El(doc, ds, "SystemCountry", S(d, "supplier_country"));
            El(doc, ds, "Short", S(d, "supplier_short"));
            El(doc, ds, "Name", S(d, "supplier_name"));
            El(doc, ds, "Type", S(d, "supplier_type"));
            El(doc, cd, "DataOperation", S(d, "data_operation"));
        }

        var fundsEl = El(doc, root, "Funds");
        var fundRows = new System.Collections.Generic.List<
            System.Collections.Generic.Dictionary<string, object?>>();
        using (var fr = Q(c, "SELECT * FROM fund WHERE document_id=@p0 "
                   + "ORDER BY fund_seq", docId))
            while (fr.Read())
                fundRows.Add(new() {
                    ["fund_seq"] = fr.GetInt32(fr.GetOrdinal("fund_seq")),
                    ["currency"] = S(fr, "currency"),
                    ["lei"] = S(fr, "lei"),
                    ["official_name"] = S(fr, "official_name"),
                    ["single_fund_flag"] = S(fr, "single_fund_flag"),
                    ["nav_date"] = S(fr, "nav_date"),
                    ["total_nav"] = D(fr, "total_nav") });

        foreach (var f in fundRows)
        {
            var ccy = (string)f["currency"]!;
            int fundSeq = (int)f["fund_seq"]!;
            var navDate = (string?)f["nav_date"];
            var fund = El(doc, fundsEl, "Fund");
            if (f["lei"] != null)
                El(doc, El(doc, fund, "Identifiers"), "LEI",
                   (string?)f["lei"]);
            El(doc, El(doc, fund, "Names"), "OfficialName",
               (string?)f["official_name"]);
            El(doc, fund, "Currency", ccy);
            if (f["single_fund_flag"] != null)
                El(doc, fund, "SingleFundFlag",
                   (string?)f["single_fund_flag"]);

            var fdd = El(doc, fund, "FundDynamicData");
            var tavp = El(doc, El(doc, fdd, "TotalAssetValues"),
                "TotalAssetValue");
            El(doc, tavp, "NavDate", navDate);
            El(doc, tavp, "TotalAssetNature", "OFFICIAL");
            var amt = El(doc, El(doc, tavp, "TotalNetAssetValue"), "Amount",
                Inv((double)f["total_nav"]!));
            amt.SetAttribute("ccy", ccy);

            var ports = El(doc, fdd, "Portfolios");
            var portRows = new System.Collections.Generic.List<(int seq,
                string? nav)>();
            using (var pr = Q(c, "SELECT portfolio_seq,nav_date FROM portfolio"
                       + " WHERE document_id=@p0 AND fund_seq=@p1 ORDER BY "
                       + "portfolio_seq", docId, fundSeq))
                while (pr.Read())
                    portRows.Add((pr.GetInt32(0),
                        pr.IsDBNull(1) ? null : pr.GetString(1)));
            foreach (var (pseq, pnav) in portRows)
            {
                var pe = El(doc, ports, "Portfolio");
                El(doc, pe, "NavDate", pnav);
                var poss = El(doc, pe, "Positions");
                var posRows = new System.Collections.Generic.List<
                    System.Collections.Generic.Dictionary<string, object?>>();
                using (var qr = Q(c, "SELECT * FROM position WHERE "
                           + "document_id=@p0 AND fund_seq=@p1 AND "
                           + "portfolio_seq=@p2 ORDER BY position_seq",
                           docId, fundSeq, pseq))
                    while (qr.Read())
                        posRows.Add(new() {
                            ["unique_id"] = S(qr, "unique_id"),
                            ["isin"] = S(qr, "isin"),
                            ["currency"] = S(qr, "currency"),
                            ["value"] = D(qr, "value_fund_ccy"),
                            ["pct"] = D(qr, "percentage"),
                            ["kind"] = S(qr, "kind"),
                            ["qty"] = Null(qr, "kind_qty")
                                ? (object?)null : D(qr, "kind_qty") });
                foreach (var p in posRows)
                {
                    var pos = El(doc, poss, "Position");
                    El(doc, pos, "UniqueID", (string?)p["unique_id"]);
                    if (p["isin"] != null)
                        El(doc, El(doc, pos, "Identifiers"), "ISIN",
                           (string?)p["isin"]);
                    if (p["currency"] != null)
                        El(doc, pos, "Currency", (string?)p["currency"]);
                    var tv = El(doc, El(doc, pos, "TotalValue"), "Amount",
                        Inv((double)p["value"]!));
                    tv.SetAttribute("ccy", ccy);
                    El(doc, pos, "TotalPercentage", Inv((double)p["pct"]!));
                    var kind = p["kind"] is string k
                        && PositionKinds.Contains(k) ? k : "Generic";
                    var ke = El(doc, pos, kind);
                    if (QtyElem.ContainsKey(kind) && p["qty"] != null)
                        El(doc, ke, QtyElem[kind], Inv((double)p["qty"]!));
                }
            }

            var scRows = new System.Collections.Generic.List<
                System.Collections.Generic.Dictionary<string, object?>>();
            using (var sr = Q(c, "SELECT * FROM share_class WHERE "
                       + "document_id=@p0 AND fund_seq=@p1 ORDER BY isin",
                       docId, fundSeq))
                while (sr.Read())
                    scRows.Add(new() {
                        ["isin"] = S(sr, "isin"),
                        ["official_name"] = S(sr, "official_name"),
                        ["currency"] = S(sr, "currency"),
                        ["nav_price"] = Null(sr, "nav_price")
                            ? (object?)null : D(sr, "nav_price"),
                        ["nav_fund_ccy"] = Null(sr, "nav_fund_ccy")
                            ? (object?)null : D(sr, "nav_fund_ccy"),
                        ["shares"] = Null(sr, "shares_outstanding")
                            ? (object?)null : D(sr, "shares_outstanding") });
            if (scRows.Count > 0)
            {
                var sce = El(doc, El(doc, fund, "SingleFund"),
                    "ShareClasses");
                foreach (var sc in scRows)
                {
                    var x = El(doc, sce, "ShareClass");
                    El(doc, El(doc, x, "Identifiers"), "ISIN",
                       (string?)sc["isin"]);
                    if (sc["official_name"] != null)
                        El(doc, El(doc, x, "Names"), "OfficialName",
                           (string?)sc["official_name"]);
                    El(doc, x, "Currency", (string?)sc["currency"]);
                    if (sc["nav_price"] != null)
                    {
                        var pr2 = El(doc, El(doc, x, "Prices"), "Price");
                        El(doc, pr2, "ActionCode", "C");
                        El(doc, pr2, "NavDate", navDate);
                        El(doc, pr2, "PriceCurrency",
                           (string?)sc["currency"]);
                        El(doc, pr2, "PriceNature", "OFFICIAL");
                        El(doc, pr2, "NavPrice",
                           Inv((double)sc["nav_price"]!));
                    }
                    if (sc["nav_fund_ccy"] != null)
                    {
                        var t2 = El(doc, El(doc, x, "TotalAssetValues"),
                            "TotalAssetValue");
                        El(doc, t2, "NavDate", navDate);
                        El(doc, t2, "TotalAssetNature", "OFFICIAL");
                        var a2 = El(doc, El(doc, t2, "TotalNetAssetValue"),
                            "Amount", Inv((double)sc["nav_fund_ccy"]!));
                        a2.SetAttribute("ccy", ccy);
                        if (sc["shares"] != null)
                            El(doc, t2, "SharesOutstanding",
                               ((double)sc["shares"]!).ToString("0",
                                   CultureInfo.InvariantCulture));
                    }
                }
            }
        }

        var assetRows = new System.Collections.Generic.List<
            System.Collections.Generic.Dictionary<string, string?>>();
        using (var ar = Q(c, "SELECT * FROM asset WHERE document_id=@p0 "
                   + "ORDER BY unique_id", docId))
            while (ar.Read())
                assetRows.Add(new() {
                    ["unique_id"] = S(ar, "unique_id"),
                    ["isin"] = S(ar, "isin"),
                    ["currency"] = S(ar, "currency"),
                    ["country"] = S(ar, "country"),
                    ["name"] = S(ar, "name"),
                    ["asset_type"] = S(ar, "asset_type") });
        if (assetRows.Count > 0)
        {
            var amd = El(doc, root, "AssetMasterData");
            foreach (var a in assetRows)
            {
                var ae = El(doc, amd, "Asset");
                El(doc, ae, "UniqueID", a["unique_id"]);
                if (a["isin"] != null)
                    El(doc, El(doc, ae, "Identifiers"), "ISIN", a["isin"]);
                El(doc, ae, "Currency", a["currency"]);
                if (a["country"] != null)
                    El(doc, ae, "Country", a["country"]);
                El(doc, ae, "Name", a["name"]);
                El(doc, ae, "AssetType", a["asset_type"]);
            }
        }

        var ws = new XmlWriterSettings
            { Indent = true, Encoding = new System.Text.UTF8Encoding(false) };
        using var w = XmlWriter.Create(outPath, ws);
        doc.Save(w);
        Console.WriteLine("wrote " + outPath);
        return 0;
    }
}

File: Database_Integration/python/export_fundsxml.py
#!/usr/bin/env python3
# =============================================================================
# EXPORT  —  relational database  ->  FundsXML file  (Python 3, SQLite).
#
# Standalone, copy-me example of ONE direction only: turning the relational
# rows back into a FundsXML document. The reverse (FundsXML -> DB) is a
# separate program, import_fundsxml.py. Over-commented as documentation.
#
# DB SCHEMA  ../ddl/schema.sql  (already populated by import_fundsxml.py).
#
# RUN
#   python3 import_fundsxml.py fx.db some.xml          # produces a doc id
#   python3 export_fundsxml.py fx.db FUNDSXML_MULTI_1 out.xml
#
# Prove the round-trip (import file vs exported file):
#   python3 ../tools/xml_equiv.py some.xml out.xml
#   XSD_Validation/cli/validate.sh <FundsXML.xsd path or release URL> out.xml
#
# DEPENDENCIES  Python stdlib `sqlite3` + `lxml`.
#
# FUNDSXML NOTES
#   * No XML namespace -> plain element names.
#   * Output is normalized to the 4.2.9 schema URL. Constants the model does
#     not store (TotalAssetNature=OFFICIAL, Price ActionCode=C / PriceNature=
#     OFFICIAL) are reproduced verbatim so the round-trip compares equal
#     (xml_equiv.py ignores the DocumentGenerated timestamp & number/whitespace
#     formatting; it is always paired with XSD validation).
#   * Rows are emitted ordered by the 1-based *_seq columns the import wrote,
#     so multiple funds/portfolios/positions come back in the original order.
# =============================================================================
import sqlite3
import sys
from datetime import datetime, timezone

from lxml import etree

SCHEMA_URL = ("https://github.com/fundsxml/schema/releases/download/"
              "4.2.9/FundsXML.xsd")
XSI = "http://www.w3.org/2001/XMLSchema-instance"
POSITION_KINDS = {"Equity", "Bond", "ShareClass", "Warrant", "Certificate",
                  "Option", "Future", "FXForward", "Swap", "Repo",
                  "RealEstate", "CallMoney", "Account", "Generic"}
QTY_ELEM = {"Equity": "Units", "Warrant": "Units", "Certificate": "Units",
            "Bond": "Nominal", "ShareClass": "Shares",
            "Option": "Contracts", "Future": "Contracts"}


def _el(parent, tag, text=None, **attrs):
    """Append a child element — the single XML-building primitive."""
    e = etree.SubElement(parent, tag)
    for k, v in attrs.items():
        e.set(k, str(v))
    if text is not None:
        e.text = str(text)
    return e


def main() -> int:
    if len(sys.argv) != 4:
        print("usage: export_fundsxml.py <db> <document_id> <out.xml>",
              file=sys.stderr)
        return 2
    db, document_id, out_path = sys.argv[1], sys.argv[2], sys.argv[3]

    con = sqlite3.connect(db)
    con.row_factory = sqlite3.Row
    d = con.execute("SELECT * FROM document WHERE document_id=?",
                     (document_id,)).fetchone()
    if d is None:
        raise SystemExit(f"no document {document_id!r} in {db}")

    root = etree.Element("FundsXML4", nsmap={"xsi": XSI})
    root.set(f"{{{XSI}}}noNamespaceSchemaLocation", SCHEMA_URL)

    cd = _el(root, "ControlData")
    _el(cd, "UniqueDocumentID", d["document_id"])
    # Regenerate the timestamp; xml_equiv.py ignores its value by design.
    _el(cd, "DocumentGenerated",
        d["generated"] or datetime.now(timezone.utc)
        .strftime("%Y-%m-%dT%H:%M:%S"))
    if d["version"]:                       # absent for FundsXML 4.0.0
        _el(cd, "Version", d["version"])
    _el(cd, "ContentDate", d["content_date"])
    ds = _el(cd, "DataSupplier")
    _el(ds, "SystemCountry", d["supplier_country"])
    _el(ds, "Short", d["supplier_short"])
    _el(ds, "Name", d["supplier_name"])
    _el(ds, "Type", d["supplier_type"])
    _el(cd, "DataOperation", d["data_operation"])

    funds_el = _el(root, "Funds")
    for f in con.execute(
            "SELECT * FROM fund WHERE document_id=? ORDER BY fund_seq",
            (document_id,)):
        ccy = f["currency"]
        fund = _el(funds_el, "Fund")
        if f["lei"]:
            _el(_el(fund, "Identifiers"), "LEI", f["lei"])
        _el(_el(fund, "Names"), "OfficialName", f["official_name"])
        _el(fund, "Currency", ccy)
        if f["single_fund_flag"]:
            _el(fund, "SingleFundFlag", f["single_fund_flag"])

        fdd = _el(fund, "FundDynamicData")
        tav = _el(_el(_el(fdd, "TotalAssetValues"), "TotalAssetValue"),
                  "NavDate", f["nav_date"]).getparent()
        _el(tav, "TotalAssetNature", "OFFICIAL")
        _el(_el(tav, "TotalNetAssetValue"), "Amount",
            f'{f["total_nav"]:.2f}', ccy=ccy)

        ports = _el(fdd, "Portfolios")
        for pf in con.execute(
                "SELECT * FROM portfolio WHERE document_id=? AND fund_seq=? "
                "ORDER BY portfolio_seq", (document_id, f["fund_seq"])):
            pe = _el(ports, "Portfolio")
            _el(pe, "NavDate", pf["nav_date"])
            poss = _el(pe, "Positions")
            for p in con.execute(
                    "SELECT * FROM position WHERE document_id=? AND fund_seq=? "
                    "AND portfolio_seq=? ORDER BY position_seq",
                    (document_id, f["fund_seq"], pf["portfolio_seq"])):
                pos = _el(poss, "Position")
                _el(pos, "UniqueID", p["unique_id"])
                if p["isin"]:
                    _el(_el(pos, "Identifiers"), "ISIN", p["isin"])
                if p["currency"]:
                    _el(pos, "Currency", p["currency"])
                _el(_el(pos, "TotalValue"), "Amount",
                    f'{p["value_fund_ccy"]:.2f}', ccy=ccy)
                _el(pos, "TotalPercentage", f'{p["percentage"]:.2f}')
                kind = p["kind"] if p["kind"] in POSITION_KINDS else "Generic"
                ke = _el(pos, kind)
                if kind in QTY_ELEM and p["kind_qty"] is not None:
                    _el(ke, QTY_ELEM[kind], f'{p["kind_qty"]:.2f}')

        scs = con.execute(
            "SELECT * FROM share_class WHERE document_id=? AND fund_seq=? "
            "ORDER BY isin", (document_id, f["fund_seq"])).fetchall()
        if scs:
            sce = _el(_el(fund, "SingleFund"), "ShareClasses")
            for sc in scs:
                x = _el(sce, "ShareClass")
                _el(_el(x, "Identifiers"), "ISIN", sc["isin"])
                if sc["official_name"]:
                    _el(_el(x, "Names"), "OfficialName", sc["official_name"])
                _el(x, "Currency", sc["currency"])
                if sc["nav_price"] is not None:
                    pr = _el(_el(x, "Prices"), "Price")
                    _el(pr, "ActionCode", "C")
                    _el(pr, "NavDate", f["nav_date"])
                    _el(pr, "PriceCurrency", sc["currency"])
                    _el(pr, "PriceNature", "OFFICIAL")
                    _el(pr, "NavPrice", f'{sc["nav_price"]:.2f}')
                if sc["nav_fund_ccy"] is not None:
                    t = _el(_el(x, "TotalAssetValues"), "TotalAssetValue")
                    _el(t, "NavDate", f["nav_date"])
                    _el(t, "TotalAssetNature", "OFFICIAL")
                    _el(_el(t, "TotalNetAssetValue"), "Amount",
                        f'{sc["nav_fund_ccy"]:.2f}', ccy=ccy)
                    if sc["shares_outstanding"] is not None:
                        _el(t, "SharesOutstanding",
                            f'{sc["shares_outstanding"]:.0f}')

    assets = con.execute(
        "SELECT * FROM asset WHERE document_id=? ORDER BY unique_id",
        (document_id,)).fetchall()
    if assets:
        amd = _el(root, "AssetMasterData")
        for a in assets:
            ae = _el(amd, "Asset")
            _el(ae, "UniqueID", a["unique_id"])
            if a["isin"]:
                _el(_el(ae, "Identifiers"), "ISIN", a["isin"])
            _el(ae, "Currency", a["currency"])
            if a["country"]:
                _el(ae, "Country", a["country"])
            _el(ae, "Name", a["name"])
            _el(ae, "AssetType", a["asset_type"])
    con.close()

    etree.ElementTree(root).write(out_path, xml_declaration=True,
                                  encoding="UTF-8", pretty_print=True)
    print("wrote", out_path)
    return 0


if __name__ == "__main__":
    sys.exit(main())
